src/OpenTelemetry/Metrics/ThreadStaticStorage.cs

```diff
@@ -49,30 +49,63 @@ internal static ThreadStaticStorage GetStorage()
         }
 
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        internal void SplitToKeysAndValues(ReadOnlySpan<KeyValuePair<string, object>> tags, out string[] tagKeys, out object[] tagValues)
+        internal void SplitToKeysAndValues(ReadOnlySpan<KeyValuePair<string, object>> tags, int tagLength, out string[] tagKeys, out object[] tagValues)
         {
-            var len = tags.Length;
+            if (tagLength <= MaxTagCacheSize)
+            {
+                tagKeys = this.tagStorage[tagLength - 1].TagKey;
+                tagValues = this.tagStorage[tagLength - 1].TagValue;
+            }
+            else
+            {
+                tagKeys = new string[tagLength];
+                tagValues = new object[tagLength];
+            }
+
+            for (var n = 0; n < tagLength; n++)
+            {
+                tagKeys[n] = tags[n].Key;
+                tagValues[n] = tags[n].Value;
+            }
+        }
 
-            if (len == 0)
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal void SplitToKeysAndValues(ReadOnlySpan<KeyValuePair<string, object>> tags, int tagLength, HashSet<string> tagKeysInteresting, out string[] tagKeys, out object[] tagValues)
+        {
+            // Iterate over tags to find the exact length.
+            int i = 0;
+            for (var n = 0; n < tagLength; n++)
             {
-                throw new InvalidOperationException("There must be atleast one tag to use ThreadStaticStorage.");
+                if (tagKeysInteresting.Contains(tags[n].Key))
+                {
+                    i++;
+                }
             }
 
-            if (len <= MaxTagCacheSize)
+            int actualLength = i;
+
+            if (actualLength <= MaxTagCacheSize)
             {
-                tagKeys = this.tagStorage[len - 1].TagKey;
-                tagValues = this.tagStorage[len - 1].TagValue;
+                tagKeys = this.tagStorage[actualLength - 1].TagKey;
+                tagValues = this.tagStorage[actualLength - 1].TagValue;
             }
             else
             {
-                tagKeys = new string[len];
-                tagValues = new object[len];
+                tagKeys = new string[actualLength];
+                tagValues = new object[actualLength];
             }
 
-            for (var n = 0; n < len; n++)
+            // Iterate again (!) to assign the actual value.
+            i = 0;
+            for (var n = 0; n < tags.Length; n++)
```

Comment:
@CodeBlanch and I discussed this offline.
Alternate to the dual iteration is to use a List with initial capacity of maxlength.. We'd be "wasting" some memory, but it'll avoid the need of multiple iterations.
Will do benchmarks and pick.